tests/test_api.py

```diff
@@ -467,6 +469,100 @@ def test_create_specification_auth(testclient):
+                "specification": json.dumps({
+                    "name": environment_name,
+                    "channels": ["main"],
+                    "dependencies": [f"pytest={version}"],
+                }),
+            },
+            timeout=10,
+        )
+        response.raise_for_status()
+        r = schema.APIPostSpecification.parse_obj(response.json())
+        assert r.status == schema.APIStatus.OK
+        build_ids.append(r.data.build_id)
+
+    # How long it takes to do a single build
+    build_delta = None
+
+    # Checks whether the builds are done (in the order they were scheduled in)
+    start = datetime.datetime.now()
+    while True:
+        # Prints the current build ids in the queue. Visually, if the server is
+        # configured to run N jobs in parallel, the queue should have N jobs
+        # less almost instantly once the first batch is done processing. After
+        # that, the queue should keep shrinking at a steady pace after each of
+        # the workers is done
+        print("build_ids", build_ids)
+
+        # Checks whether the time limit is reached
+        now = datetime.datetime.now()
+        delta_seconds = (now - start).total_seconds()
+        if delta_seconds > limit_seconds:
+            break
+
+        # Measures how long it takes to do a single build
+        if build_delta is None and len(build_ids) < num_builds:
+            build_delta = delta_seconds
+
+        # Gets the oldest build in the queue as it's the one that's most likely
+        # to be done
+        try:
+            build_id = build_ids.popleft()
+        except IndexError:
+            break
+
+        # Checks the status
+        response = testclient.get(f"api/v1/build/{build_id}", timeout=10)
+        response.raise_for_status()
+        r = schema.APIGetBuild.parse_obj(response.json())
+        assert r.status == schema.APIStatus.OK
+        assert r.data.specification.name == environment_name
+
+        # Exits immediately on failure
+        assert r.data.status != 'FAILED'
+
+        # If not done, adds the id back to the end of the queue
+        if r.data.status != 'COMPLETED':
+            build_ids.append(build_id)
+
+        # Adds a small delay to avoid making too many requests too fast. The
+        # build takes significantly longer than 1 second, so this shouldn't
+        # impact the measurements
+        time.sleep(1)
+
+    # Measures how long it took to run the loop
+    loop_delta = (datetime.datetime.now() - start).total_seconds()
+    print("build_delta", build_delta)
+    print("loop_delta", loop_delta)
+
+    # If there are jobs in the queue, the loop didn't complete in the allocated
+    # time. So something went wrong, like a build getting stuck or parallel
+    # builds not working
+    assert len(build_ids) == 0
+
+    # If all jobs are done twice as fast as they would do sequentially, then
+    # parallel builds are working
+    assert loop_delta < build_delta * (num_builds / 2)
```

Comment:
This is a flawed assumption. A different check is needed. It doesn't work because there's some additional time needed before the first build actually starts executing. Subsequent builds (from build 2+) are much faster.

With concurrency=4, I get:
```
build_delta 178.358115
loop_delta 411.383843
```

(Note: my test VM only has 2 CPUs, exposed, which explains why this is only 2x faster than with concurrency=1.)

With concurrency=1, I get:
```
build_delta 427.717276
loop_delta 784.000384
```

Since this is an integration test, I cannot just measure against different concurrency parameters in the test. I think I might try comparing the deltas between subsequent jobs (1 vs 2, 2 vs 3, etc.), with the assumption that with concurrency=1 these will be roughly the same, and with concurrency=4 these will be done in batches.